vpr/src/route/route_tree.h

```diff
@@ -0,0 +1,519 @@
+    RouteTreeNode* _next = nullptr;
+
+    /** Having a doubly linked list helps maintain the ordering of nodes */
+    RouteTreeNode* _prev = nullptr;
+
+    /** Here is the awkward part: these two pointers can become a union. 
+     * 1. If there is a next sibling, _subtree_end is equal to it: when we finish walking the subtree we arrive at the next sibling.
+     * 2. If there is not, _subtree_end is equal to the parent's _subtree_end.
+     * So there is never a case where _next_sibling and _subtree_end are both valid and pointing at different things.
+     * Then, the only question is where to stop when walking _next_sibling ptrs, since the final sibling won't point at nullptr.
+     * But it will point at parent's _subtree_end, so we can use that as the limiting case */
+    union {
+        /* Ptr to "next sibling": used when traversing child nodes */
+        RouteTreeNode* _next_sibling = nullptr;
+        /* Where does the subtree under this node end in the flattened linked list? Needed when recursively iterating */
+        RouteTreeNode* _subtree_end;
+    };
+};
+
+/** fwd definition for compatibility class in old_traceback.h */
+class TracebackCompat;
+
+/**
+ * @brief Top level route tree used in timing analysis and keeping routing state.
+ *
+ * Contains the root node and a lookup from RRNodeIds to RouteTreeNode&s in the tree. */
+class RouteTree {
+    friend class TracebackCompat;
+
+  public:
+    RouteTree() = delete;
+    RouteTree(const RouteTree&);
+    RouteTree(RouteTree&&);
+    RouteTree& operator=(const RouteTree&);
+    RouteTree& operator=(RouteTree&&);
+
+    /** Return a RouteTree initialized to inode. */
+    RouteTree(RRNodeId inode);
+    /** Return a RouteTree initialized to the source of nets[inet]. */
+    RouteTree(ParentNetId inet);
+
+    ~RouteTree() {
+        free_list(_root);
+    }
+
+    /** Add the most recently finished wire segment to the routing tree, and
+     * update the Tdel, etc. numbers for the rest of the routing tree. hptr
+     * is the heap pointer of the SINK that was reached, and target_net_pin_index
+     * is the net pin index corresponding to the SINK that was reached. This routine
+     * returns a tuple: RouteTreeNode of the branch it adds to the route tree and
+     * RouteTreeNode of the SINK it adds to the routing. */
+    std::tuple<vtr::optional<const RouteTreeNode&>, vtr::optional<const RouteTreeNode&>>
+    update_from_heap(t_heap* hptr, int target_net_pin_index, SpatialRouteTreeLookup* spatial_rt_lookup, bool is_flat);
+
+    /** Reload timing values (R_upstream, C_downstream, Tdel).
+     * Can take a RouteTreeNode& to do an incremental update.
+     * Note that update_from_heap already does this, but prune() doesn't */
+    void reload_timing(vtr::optional<RouteTreeNode&> from_node = vtr::nullopt);
+
+    /** Get the RouteTreeNode corresponding to the RRNodeId. Returns nullopt if not found. */
```

Comment:
If this rr_node is in the tree multiple times, it returns the last one that was added to the tree.